File: packages/mpc-wallet-sdk/src/chains/solana.ts
/**
 * Solana chain adapter
 */

import type { Balance, TxHash } from '../types';

/**
 * Solana chain configuration
 */
export interface SolanaChainConfig {
  /** Chain name */
  name: string;
  /** RPC URLs (with failover) */
  rpcUrls: string[];
  /** Commitment level */
  commitment: 'processed' | 'confirmed' | 'finalized';
  /** Explorer URL */
  explorerUrl?: string;
}

/**
 * Pre-configured Solana networks
 */
export const SolanaNetworks = {
  MAINNET: {
    name: 'Solana Mainnet',
    rpcUrls: ['https://api.mainnet-beta.solana.com'],
    commitment: 'confirmed' as const,
    explorerUrl: 'https://explorer.solana.com',
  },
  DEVNET: {
    name: 'Solana Devnet',
    rpcUrls: ['https://api.devnet.solana.com'],
    commitment: 'confirmed' as const,
    explorerUrl: 'https://explorer.solana.com?cluster=devnet',
  },
  TESTNET: {
    name: 'Solana Testnet',
    rpcUrls: ['https://api.testnet.solana.com'],
    commitment: 'confirmed' as const,
    explorerUrl: 'https://explorer.solana.com?cluster=testnet',
  },
} as const;

/**
 * Transaction parameters for Solana
 */
export interface SolanaTxParams {
  /** Sender public key (base58) */
  from: string;
  /** Recipient public key (base58) */
  to: string;
  /** Amount in lamports */
  amount: bigint;
  /** Recent blockhash (will be fetched if not provided) */
  recentBlockhash?: string;
  /** Priority fee in microlamports per compute unit */
  priorityFee?: number;
}

/**
 * Unsigned Solana transaction
 */
export interface UnsignedSolanaTx {
  /** Serialized message for signing */
  message: Uint8Array;
  /** Recent blockhash used */
  recentBlockhash: string;
  /** Human-readable summary */
  summary: {
    from: string;
    to: string;
    amount: string;
    estimatedFee: string;
  };
}

/**
 * Solana chain adapter
 *
 * @example
 * ```typescript
 * const adapter = new SolanaAdapter(SolanaNetworks.MAINNET);
 *
 * // Get balance
 * const balance = await adapter.getBalance('...');
 *
 * // Build transaction
 * const unsignedTx = await adapter.buildTransaction({
 *   from: '...',
 *   to: '...',
 *   amount: 1000000000n, // 1 SOL
 * });
 * ```
 */
export class SolanaAdapter {
  private config: SolanaChainConfig;
  private currentRpcIndex = 0;

  constructor(config: SolanaChainConfig) {
    this.config = config;
  }

  /** Get network name */
  get name(): string {
    return this.config.name;
  }

  /** Native currency symbol */
  get symbol(): string {
    return 'SOL';
  }

  /** Native currency decimals */
  get decimals(): number {
    return 9;
  }

  /**
   * Get SOL balance for an address
   */
  async getBalance(address: string): Promise<Balance> {
    const result = await this.rpcCall<{ value: number }>('getBalance', [
      address,
      { commitment: this.config.commitment },
    ]);

    const rawValue = BigInt(result.value);
    const formatted = this.formatLamports(rawValue);

    return {
      raw: rawValue.toString(),
      formatted,
      symbol: 'SOL',
      decimals: 9,
    };
  }

  /**
   * Get recent blockhash
   */
  async getRecentBlockhash(): Promise<string> {
    const result = await this.rpcCall<{
      value: { blockhash: string; lastValidBlockHeight: number };
    }>('getLatestBlockhash', [{ commitment: this.config.commitment }]);

    return result.value.blockhash;
  }

  /**
   * Get priority fee estimate
   */
  async getPriorityFee(): Promise<number> {
    try {
      const result = await this.rpcCall<{
        prioritizationFeeEstimate: { priorityFeeEstimate: number };
      }>('getRecentPrioritizationFees', []);

      return result.prioritizationFeeEstimate?.priorityFeeEstimate ?? 1000;
    } catch {
      return 1000; // Default to 1000 microlamports
    }
  }

  /**
   * Build an unsigned transaction
   */
  async buildTransaction(params: SolanaTxParams): Promise<UnsignedSolanaTx> {
    const recentBlockhash =
      params.recentBlockhash ?? (await this.getRecentBlockhash());

    // Build a simple transfer instruction
    // In production, this would use proper Solana SDK serialization
    const message = this.buildTransferMessage(
      params.from,
      params.to,
      params.amount,
      recentBlockhash
    );

    // Base fee + priority fee
    const baseFee = 5000n; // 5000 lamports
    const priorityFee = BigInt(params.priorityFee ?? 1000) * 200n; // Estimate 200 CU
    const estimatedFee = baseFee + priorityFee;

    return {
      message,
      recentBlockhash,
      summary: {
        from: params.from,
        to: params.to,
        amount: this.formatLamports(params.amount),
        estimatedFee: this.formatLamports(estimatedFee),
      },
    };
  }

  /**
   * Finalize a transaction with signature
   */
  finalizeTransaction(
    unsignedTx: UnsignedSolanaTx,
    signature: Uint8Array
  ): Uint8Array {
    // Combine message with signature
    // Solana transaction format: [signature_count, signatures..., message]
    const signatureCount = 1;
    const tx = new Uint8Array(1 + 64 + unsignedTx.message.length);
    tx[0] = signatureCount;
    tx.set(signature.slice(0, 64), 1);
    tx.set(unsignedTx.message, 65);
    return tx;
  }

  /**
   * Broadcast a signed transaction
   */
  async broadcast(signedTx: Uint8Array): Promise<TxHash> {
    const txBase64 = btoa(String.fromCharCode(...signedTx));

    const result = await this.rpcCall<string>('sendTransaction', [
      txBase64,
      {
        encoding: 'base64',
        preflightCommitment: this.config.commitment,
      },
    ]);

    return {
      hash: result,
      explorerUrl: this.config.explorerUrl
        ? `${this.config.explorerUrl}/tx/${result}`
        : undefined,
    };
  }

  /**
   * Wait for transaction confirmation
   */
  async waitForConfirmation(
    signature: string,
    timeoutMs = 60000
  ): Promise<{ confirmed: boolean; slot?: number }> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeoutMs) {
      try {
        const result = await this.rpcCall<{
          value: { confirmationStatus: string; slot: number } | null;
        }>('getSignatureStatuses', [[signature]]);

        const status = result.value;
        if (status) {
          const confirmationStatus = status.confirmationStatus;
          if (
            confirmationStatus === 'confirmed' ||
            confirmationStatus === 'finalized'
          ) {
            return { confirmed: true, slot: status.slot };
          }
        }
      } catch {
        // Status not available yet
      }

      await sleep(2000);
    }

    return { confirmed: false };
  }

  /**
   * Check if an address is valid (base58)
   */
  isValidAddress(address: string): boolean {
    // Base58 alphabet check
    return /^[1-9A-HJ-NP-Za-km-z]{32,44}$/.test(address);
  }

  /**
   * Get explorer URL for a transaction
   */
  getExplorerTxUrl(signature: string): string | undefined {
    if (!this.config.explorerUrl) return undefined;
    return `${this.config.explorerUrl}/tx/${signature}`;
  }

  /**
   * Get explorer URL for an address
   */
  getExplorerAddressUrl(address: string): string | undefined {
    if (!this.config.explorerUrl) return undefined;
    return `${this.config.explorerUrl}/address/${address}`;
  }

  // ============================================================================
  // Private Methods
  // ============================================================================

  private async rpcCall<T>(method: string, params: unknown[]): Promise<T> {
    const errors: Error[] = [];

    for (let i = 0; i < this.config.rpcUrls.length; i++) {
      const rpcUrl = this.config.rpcUrls[this.currentRpcIndex];

      try {
        const response = await fetch(rpcUrl, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            jsonrpc: '2.0',
            method,
            params,
            id: Date.now(),
          }),
        });

        const data = await response.json();

        if (data.error) {
          throw new Error(data.error.message || 'RPC error');
        }

        return data.result;
      } catch (error) {
        errors.push(error as Error);
        this.currentRpcIndex =
          (this.currentRpcIndex + 1) % this.config.rpcUrls.length;
      }
    }

    throw new Error(
      `All RPC endpoints failed: ${errors.map((e) => e.message).join(', ')}`
    );
  }

  private buildTransferMessage(
    from: string,
    to: string,
    amount: bigint,
    recentBlockhash: string
  ): Uint8Array {
    // Simplified message building
    // In production, use proper Solana SDK
    const message = new TextEncoder().encode(
      JSON.stringify({
        from,
        to,
        amount: amount.toString(),
        recentBlockhash,
        programId: '11111111111111111111111111111111', // System program
      })
    );
    return message;
  }

  private formatLamports(lamports: bigint): string {
    const sol = Number(lamports) / 1e9;
    return `${sol.toFixed(9).replace(/\.?0+$/, '')} SOL`;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: packages/mpc-wallet-python/src/mpc_wallet/chains/solana.py
"""Solana chain adapter."""

from dataclasses import dataclass
from typing import Any
import asyncio
import httpx

from ..types import Balance, TxHash


@dataclass
class SolanaChainConfig:
    """Solana chain configuration."""

    name: str
    rpc_urls: list[str]
    commitment: str = "confirmed"
    explorer_url: str | None = None


# Pre-configured Solana networks
SolanaNetworks = {
    "MAINNET": SolanaChainConfig(
        name="Solana Mainnet",
        rpc_urls=["https://api.mainnet-beta.solana.com"],
        explorer_url="https://explorer.solana.com",
    ),
    "DEVNET": SolanaChainConfig(
        name="Solana Devnet",
        rpc_urls=["https://api.devnet.solana.com"],
        explorer_url="https://explorer.solana.com?cluster=devnet",
    ),
    "TESTNET": SolanaChainConfig(
        name="Solana Testnet",
        rpc_urls=["https://api.testnet.solana.com"],
        explorer_url="https://explorer.solana.com?cluster=testnet",
    ),
}


@dataclass
class SolanaTxParams:
    """Transaction parameters for Solana."""

    from_address: str  # Base58 pubkey
    to: str  # Base58 pubkey
    amount: int  # In lamports
    recent_blockhash: str | None = None
    priority_fee: int | None = None


@dataclass
class UnsignedSolanaTx:
    """Unsigned Solana transaction."""

    message: bytes
    recent_blockhash: str
    summary: dict[str, str]


class SolanaAdapter:
    """
    Solana chain adapter.

    Example:
        >>> adapter = SolanaAdapter(SolanaNetworks["MAINNET"])
        >>>
        >>> # Get balance
        >>> balance = await adapter.get_balance("...")
        >>>
        >>> # Build transaction
        >>> unsigned_tx = await adapter.build_transaction(SolanaTxParams(
        ...     from_address="...",
        ...     to="...",
        ...     amount=1_000_000_000,  # 1 SOL
        ... ))
    """

    def __init__(self, config: SolanaChainConfig) -> None:
        self._config = config
        self._current_rpc_index = 0

    @property
    def name(self) -> str:
        """Get network name."""
        return self._config.name

    @property
    def symbol(self) -> str:
        """Get native currency symbol."""
        return "SOL"

    @property
    def decimals(self) -> int:
        """Get native currency decimals."""
        return 9

    async def get_balance(self, address: str) -> Balance:
        """Get SOL balance for an address."""
        result = await self._rpc_call(
            "getBalance",
            [address, {"commitment": self._config.commitment}],
        )

        raw_value = result["value"]
        formatted = self._format_lamports(raw_value)

        return Balance(
            raw=str(raw_value),
            formatted=formatted,
            symbol="SOL",
            decimals=9,
        )

    async def get_recent_blockhash(self) -> str:
        """Get recent blockhash."""
        result = await self._rpc_call(
            "getLatestBlockhash",
            [{"commitment": self._config.commitment}],
        )
        return result["value"]["blockhash"]

    async def build_transaction(self, params: SolanaTxParams) -> UnsignedSolanaTx:
        """Build an unsigned transaction."""
        import json

        recent_blockhash = params.recent_blockhash
        if not recent_blockhash:
            recent_blockhash = await self.get_recent_blockhash()

        # Build a simple transfer message (simplified)
        message = json.dumps({
            "from": params.from_address,
            "to": params.to,
            "amount": params.amount,
            "recentBlockhash": recent_blockhash,
            "programId": "11111111111111111111111111111111",
        }).encode()

        # Estimate fee
        base_fee = 5000  # 5000 lamports
        priority_fee = (params.priority_fee or 1000) * 200
        estimated_fee = base_fee + priority_fee

        return UnsignedSolanaTx(
            message=message,
            recent_blockhash=recent_blockhash,
            summary={
                "from": params.from_address,
                "to": params.to,
                "amount": self._format_lamports(params.amount),
                "estimated_fee": self._format_lamports(estimated_fee),
            },
        )

    def finalize_transaction(
        self, unsigned_tx: UnsignedSolanaTx, signature: bytes
    ) -> bytes:
        """Finalize a transaction with signature."""
        # Solana transaction format: [sig_count, signatures..., message]
        tx = bytes([1]) + signature[:64] + unsigned_tx.message
        return tx

    async def broadcast(self, signed_tx: bytes) -> TxHash:
        """Broadcast a signed transaction."""
        import base64

        tx_base64 = base64.b64encode(signed_tx).decode()

        result = await self._rpc_call(
            "sendTransaction",
            [tx_base64, {"encoding": "base64", "preflightCommitment": self._config.commitment}],
        )

        explorer_url = None
        if self._config.explorer_url:
            explorer_url = f"{self._config.explorer_url}/tx/{result}"

        return TxHash(hash=result, explorer_url=explorer_url)

    async def wait_for_confirmation(
        self, signature: str, timeout_secs: int = 60
    ) -> dict[str, Any]:
        """Wait for transaction confirmation."""
        import time

        start_time = time.time()

        while time.time() - start_time < timeout_secs:
            try:
                result = await self._rpc_call("getSignatureStatuses", [[signature]])
                status = result["value"][0]
                if status:
                    confirmation = status.get("confirmationStatus", "")
                    if confirmation in ("confirmed", "finalized"):
                        return {
                            "confirmed": True,
                            "slot": status.get("slot"),
                        }
            except Exception:
                pass

            await asyncio.sleep(2)

        return {"confirmed": False}

    def is_valid_address(self, address: str) -> bool:
        """Check if an address is valid (base58)."""
        import re
        return bool(re.match(r"^[1-9A-HJ-NP-Za-km-z]{32,44}$", address))

    def get_explorer_tx_url(self, signature: str) -> str | None:
        """Get explorer URL for a transaction."""
        if not self._config.explorer_url:
            return None
        return f"{self._config.explorer_url}/tx/{signature}"

    def get_explorer_address_url(self, address: str) -> str | None:
        """Get explorer URL for an address."""
        if not self._config.explorer_url:
            return None
        return f"{self._config.explorer_url}/address/{address}"

    async def _rpc_call(self, method: str, params: list[Any]) -> Any:
        """Make an RPC call with failover."""
        errors: list[str] = []

        async with httpx.AsyncClient(timeout=30.0) as client:
            for _ in range(len(self._config.rpc_urls)):
                rpc_url = self._config.rpc_urls[self._current_rpc_index]

                try:
                    response = await client.post(
                        rpc_url,
                        json={
                            "jsonrpc": "2.0",
                            "method": method,
                            "params": params,
                            "id": 1,
                        },
                    )
                    data = response.json()

                    if "error" in data:
                        raise Exception(data["error"].get("message", "RPC error"))

                    return data["result"]

                except Exception as e:
                    errors.append(str(e))
                    self._current_rpc_index = (
                        self._current_rpc_index + 1
                    ) % len(self._config.rpc_urls)

        raise Exception(f"All RPC endpoints failed: {', '.join(errors)}")

    def _format_lamports(self, lamports: int) -> str:
        """Format lamports as SOL."""
        sol = lamports / 1e9
        return f"{sol:.9f}".rstrip("0").rstrip(".") + " SOL"
